src/library/dao/directorydao.cpp

```diff
@@ -217,8 +217,10 @@ std::pair<DirectoryDAO::RelocateResult, QList<RelocatedTrack>> DirectoryDAO::rel
     // Appending '/' is required to disambiguate files from parent
     // directories, e.g. "a/b.mp3" and "a/b/c.mp3" where "a/b" would
     // match both instead of only files in the parent directory "a/b/".
-    DEBUG_ASSERT(!oldDirectory.endsWith('/'));
-    const QString oldDirectoryPrefix = oldDirectory + '/';
+    QString oldDirectoryPrefix = oldDirectory;
+    if (!oldDirectory.endsWith('/')) {
+        oldDirectoryPrefix += '/';
+    }
```

Comment:
I think the trailing slash came up much earlier, the directory stored in the db seems to have it. Relinking seems to be the only place where this came up.

If we don't want trailing slashes, we should probably assert this when adding the directory? (and maybe here too as a sanity check)

Fwiw, double slashes are fine in Unix and simply ignored.

Reply:
Off top of my head, I think we must pass the old directory verbatim as it is stored in the db (including any trailing slash), otherwise the query won't update it, so we'd have to update this elsewhere one way or the other.

Not sure how double slashes could come up here, if we check for the trailing slash before adding it, then the only scenario where this would come up is if we would already pass in such a string. The debug assert only prevents this for the trailing case and doesn't for check intermediate //s (though since Unix is fine with them, it probably wouldn't have been an issue either way)

Reply:
AFAICT we had no issues with trailing / when relocating (after the latest reloc fixes), which indicates that assert is fine.

IMHO we should rather check where trailing / comes from in the first place (fwiw I don't have them in my `directories` table) and/or make otherwise sure we don't have // 

Do you see trailing slashes also in win/macOS/linux databases / preferences?

Reply:
On macOS/Linux I don't have any trailing slashes in my db either. I think it's either the iOS file chooser or Qt's home path (which effectively delegates to the iOS API). Just stripping any trailing slashes when adding the directory might be the best solution? I'd be hesitant to do it in the iOS-specific code, since we'd still bake the assumption that other platforms don't return any trailing slashes (which might hold, but feels a bit fragile).

Reply:
Ah, hm, we don't guard against trailing slashes in the new directory either, during relocation

Reply:
True.
I think we should fix that in 2.4. For example add something like
`while (newDirString.endsWith('/') { newDirStr.chop(1); }`

FWIW I also just figured that
* `QFileInfo::absoluteFilePath()` may return a path **with** a trailing slash if the string for used for construction had one.
used by [`mixxx::FileInfo::location()`](https://github.com/mixxxdj/mixxx/blob/17edb4e5f0554a2b3a0c5c5952d4468209b1ca87/src/util/fileinfo.h#L115-L117) which is used by `DirectoryDAO::addDirectory()`, and `DirectoryDAO::relocateDirectory()` for verifying the QString newDir
* otoh, `QFileInfo::absolutePath()` returns the directory part only (**no trailing slash**)
used by `mixxx::FileInfo::locationPath()`
(just verified on Linux with Qt 6.2.3)

Another finding:
[QFileInfo::absoluteFilePath](https://doc.qt.io/qt-6/qfileinfo.html#absoluteFilePath) will not have trailing slashes, if... ([Qt 6 documentaion](https://doc.qt.io/qt-6/qfileinfo.html#details)):
> QFileInfo objects referring to directories created by Qt classes will not have a trailing directory separator '/'.

but we just pass the _strings_ from the db around :\

___
About the origin of trailing / in music dirs:
adding a new dir works like this:
1. DlgPrefLibrary::slotAddDir() gets a **QString** from QFileDialog::getExistingDirectory()
2. Library::requestAddDir() creates a **mixxx::FileInfo** from that
3. TrackCollectionManager::addDirectory()  -> TrackCollection::addDirectory
4. DirectoryDAO::addDirectory() turns this into a **QString** with `mixxx::FileInfo::canonicalLocation()`
(which calls `QFileInfo::canonicalFilePath()`)
which does **not** have a trailing / even if I append one manually in DlgPrefLibrary::slotAddDir()
(also tested with Qt 6.2.3 on Linux)

Conclusion: the trailing slash may only stem from relocating?

Reply:
Tbh is it really that much of an issue if we store directories which may or may not have a trailing slash? I feel like documenting and upholding this invariant everywhere (across whichever platform APIs we encounter) is probably going to be fragile, one way or another.

Trailing slashes seems to work fine for almost everything else and if we really need a path with (or without) the slash, then adding or removing it should be easy where needed

Reply:
But yeah, if it only stems from relocating, maybe we could just call `QFileInfo::absolutePath` or `mixxx::FileInfo::canonicalPath` on the new directory to strip it?

Reply:
Well, I'm really glad relocating seems to work now, even cross-OS.
(failed (cross-OS ) relocation means loosing the library, and debugging that (remotely in the forums) is PITA)

Therefore I'd really love to play it safe and not risk any regressions, i.e. prevent any trailing / or //.
(didn't check what might happen if we for example relocate to Windows from iOS which already did sandbox relocation (adds /))
The only place where we need / is for finding the old track(location)s.

So, yes, this should do the trick (IIUC :zany_face: )
> maybe we could just call QFileInfo::absolutePath or mixxx::FileInfo::canonicalPath on the new directory to strip it?

Reply:
Hm, unfortunately relocation does not seem to be the only place where this could come up, adding a directory already produces a trailing slash...

https://github.com/user-attachments/assets/e57f3ced-1b37-4f44-be2b-4b78116321ee

Reply:
Indeed. Somehow I screwed up the test 
> 1. DlgPrefLibrary::slotAddDir() gets a **QString** from QFileDialog::getExistingDirectory()
> 2. Library::requestAddDir() creates a **mixxx::FileInfo** from that
> 3. TrackCollectionManager::addDirectory()  -> TrackCollection::addDirectory
> 4. DirectoryDAO::addDirectory() turns this into a **QString** with `mixxx::FileInfo::canonicalLocation()`
> (which calls `QFileInfo::canonicalFilePath()`)
> which does **not** have a trailing / even if I append one manually in DlgPrefLibrary::slotAddDir()
> (also tested with Qt 6.2.3 on Linux)

for **comparisson with existing directories**.
But for actually **adding the directory** it uses `mixxx::FileInfo::location()` which **can have** a trailing / 

So, finally, my conclusion:
* add directory: since `QFileDialog::getExistingDirectory()` may return a path with /, let's use `newCanonicalLocation` here
https://github.com/mixxxdj/mixxx/blob/f4785f4720bc244fe51a0fdd1a1e9396d9b1e4c8/src/library/dao/directorydao.cpp#L125-L127
* relocate dir: let's make _somehow_¹ sure that `oldDirectoryPrefix` ends with **one /**
https://github.com/mixxxdj/mixxx/blob/f4785f4720bc244fe51a0fdd1a1e9396d9b1e4c8/src/library/dao/directorydao.cpp#L221
personally I don't mind whether that is a smart regex check/replacement or QString operation.

What do you think?

___
Ideally we'd have no / in the `directories` table.
Shall we add a cleanup method that is called each time someone calls `DirectoryDAO::getRootDirStrings()` or `DirectoryDAO::loadAllDirectories()`?